### DISCUSSION TITLE
TARGETARCH/TARGETPLATFORM scope and behaviour

### DISCUSSION THREAD

**Discussion by droopy4096** [docker/buildx#1975] [Q&A]
It seems that scope for "automatic" `TARGETARCH` and `TARGETPLATFORM` are no well defined. Consider example:

```Dockerfile
FROM alpine
ARG TARGETPLATFORM
RUN echo "I'm building for $TARGETPLATFORM" > /var/tmp/platform
```

as per docs - it works as expected

however

```Dockerfile
FROM alpine
RUN echo "I'm building for $TARGETPLATFORM" > /var/tmp/platform
```

does not - producing "empty" output

Yet, when I use it within `FROM ...` statement it seems to work

```Dockerfile
FROM --platform=${TARGETPLATFORM} alpine
RUN uname -m > /var/tmp/platform
```

to take it further I've tried

```Dockerfile
FROM --platform="" alpine
RUN uname -m > /var/tmp/platform
```

and it works just fine with default cli options. So I ended up using:

```shell
docker buildx build --platform=linux/amd64 -f Dockerfile -t my-my .
```

with previous Dockerfile and that did result in expected:

```
0.193 standard_init_linux.go:219: exec user process caused: exec format error
```

in other words confirming that automatic vars like `$TARGETPLATFORM` have "odd" scoping:

* they are available immediately for `FROM ...` statements
* they are *not* available for each stage execution unless explicitly requested via `ARG`. 

:question: Question:

Moving forward should we *always* use `ARG TARGETARCH` and such to declare intent to use that automatic variable or is this an intended behaviour and `FROM ...` will receive automatic variable without declaration yet anywhere else there needs to be explicit declaration?
